src/globus_sdk/transport/retry.py

```diff
@@ -0,0 +1,254 @@
+    a RetryContext to evaluate. Once a hook has determined that the request should or
+    should not be retried, or all hooks have been consulted, the policy returns a
+    result.
+
+    The RetryPolicy is also responsible for determining the amount of backoff between
+    retries, and executing the sleeps between retries itself.
+
+    :param backoff: A function which determines how long to sleep between calls based on
+        the RetryContext. Defaults to expontential backoff with jitter based on the
+        context ``attempt`` number.
+    :type backoff: callable
+    :param checks: A list of initial checks for the policy. Any hooks registered,
+        including the default hooks, will run after these checks.
+    :type checks: list of callables
+    :param max_sleep: The maximum sleep time between retries (in seconds). If the
+        computed sleep time or the backoff requested by a retry check exceeds this
+        value, this amount of time will be used instead
+    :type max_sleep: int
+    :param max_retries: The maximum number of retries allowed by this policy. This is
+        checked by ``default_check_max_retries_exceeded`` to see if a request should
+        stop retrying.
+    :type max_retries: int
+    """
+
+    #: status codes for responses which may have a Retry-After header
+    RETRY_AFTER_STATUS_CODES = (429, 503)
+    #: status codes for error responses which should generally be retried
+    TRANSIENT_ERROR_STATUS_CODES = (429, 500, 502, 503, 504)
+    #: status codes indicating that authorization info was missing or expired
+    EXPIRED_AUTHORIZATION_STATUS_CODES = (401,)
+
+    def __init__(
+        self,
+        *,
+        backoff: typing.Callable[[RetryContext], float] = _exponential_backoff,
+        checks: typing.Optional[typing.List[RetryCheck]] = None,
+        max_sleep: int = 10,
+        max_retries: int = 5,
+    ):
+        self.backoff = backoff
+        self.max_sleep = max_sleep
+        self.max_retries = max_retries
+        # copy
+        self.checks = list(checks if checks else [])
+        # register internal checks
+        self.register_default_checks()
+
+    def compute_delay(self, ctx: RetryContext) -> float:
+        """
+        Given a retry context, compute the amount of time to sleep.
+        This is always the minimum of the backoff (run on the context) and the
+        ``max_sleep``.
+        """
+        delay = self.backoff(ctx)
+        return min(delay, self.max_sleep)
+
+    def should_retry(self, context: RetryContext) -> bool:
+        """
+        Determine whether or not a request should retry by consulting all registered
+        checks.
+        """
+        for check in self.checks:
+            result = check(context)
+            if result is None:
+                continue
+            elif result is False:
+                return False
+            else:
+                time.sleep(self.compute_delay(context))
+                return True
```

Comment:
Wouldn't this make this method sensitive to the order the checks are run? I would picture the logic I expect as the following:
```suggestion
        return not any(check(context) is False for check in self.checks)
```
so perhaps the first check says to retry, but a check towards the end determines there should be no retry.

Reply:
I would also tend towards keeping the side effect `time.sleep(...)` out of a `.should_retry()` method—I would assume the caller of `should_retry` is responsible for actually doing the waiting.

Reply:
(Also `should_retry` to me sounds innocuous and non-side-effectful.)

Reply:
They are order-sensitive today, so I don't think contracting like this works unless we restructure more dramatically. Mainly, the issue is that checks can have side-effects. In particular, the authorization one, but also the Retry-After one needs a way to pass back the parsed header value (which right now is done by mutating the context object).

I'm inclined to stick with the current approach for the purposes of getting a first version merged, but I'd be extremely happy to revisit and try to improve it.  _Ideally_, checks would be forbidden from having side-effects. I tried a version at some point (not sure if it's in the commit history or not) which had a `RetryCheckResult` type for passing back true/false and a float for seconds of delay. But I think doing something like that will require changes to the authorizers too (I'm not sure exactly what -- that's where I drew the line with that experiment).